src/app/codegen-data-model/CodegenDataModel_Write.cpp

```diff
@@ -0,0 +1,367 @@
+}
+
+/// Metadata of what a ember/pascal short string means (prepended by a u8 length)
+struct ShortPascalString
+{
+    using LengthType                        = uint8_t;
+    static constexpr LengthType kNullLength = 0xFF;
+
+    static void SetLength(uint8_t * buffer, LengthType value) { *buffer = value; }
+};
+
+/// Metadata of what a ember/pascal LONG string means (prepended by a u16 length)
+struct LongPascalString
+{
+    using LengthType                        = uint16_t;
+    static constexpr LengthType kNullLength = 0xFFFF;
+
+    // Encoding for ember string lengths is little-endian (see ember-strings.cpp)
+    static void SetLength(uint8_t * buffer, LengthType value) { Encoding::LittleEndian::Put16(buffer, value); }
+};
+
+// ember assumptions ... should just work
+static_assert(sizeof(ShortPascalString::LengthType) == 1);
+static_assert(sizeof(LongPascalString::LengthType) == 2);
+
+/// Encode the value stored in 'decoder' into an ember format string 'out'
+///
+/// The value encoded will be of type T (e.g. CharSpan or ByteSpan) and it will be encoded
+/// via the given ENCODING (i.e. ShortPascalString or LongPascalString)
+///
+/// isNullable defines if the value of NULL is allowed to be encoded.
+template <typename T, class ENCODING>
+CHIP_ERROR DecodeStringLikeIntoEmberBuffer(AttributeValueDecoder decoder, bool isNullable, MutableByteSpan out)
+{
+    T workingValue;
+
+    if (isNullable)
+    {
+        typename DataModel::Nullable<T> nullableWorkingValue;
+        ReturnErrorOnFailure(decoder.Decode(nullableWorkingValue));
+
+        if (nullableWorkingValue.IsNull())
+        {
+            VerifyOrReturnError(out.size() >= sizeof(typename ENCODING::LengthType), CHIP_ERROR_BUFFER_TOO_SMALL);
+            ENCODING::SetLength(out.data(), ENCODING::kNullLength);
+            return CHIP_NO_ERROR;
+        }
+
+        // continue encoding non-null value
+        workingValue = nullableWorkingValue.Value();
+    }
+    else
+    {
+        ReturnErrorOnFailure(decoder.Decode(workingValue));
+    }
+
+    typename ENCODING::LengthType len = static_cast<typename ENCODING::LengthType>(workingValue.size());
+    VerifyOrReturnError(out.size() >= sizeof(len) + len, CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    uint8_t * output_buffer = out.data();
+
+    ENCODING::SetLength(output_buffer, len);
+    output_buffer += sizeof(len);
+
+    memcpy(output_buffer, workingValue.data(), workingValue.size());
+
+    return CHIP_NO_ERROR;
+}
+
+/// Encodes a numeric data value of type T from the given ember-encoded buffer `data`.
+///
+/// isNullable defines if the value of NULL is allowed to be encoded.
+template <typename T>
+CHIP_ERROR DecodeIntoEmberBuffer(AttributeValueDecoder & decoder, bool isNullable, MutableByteSpan out)
+{
+    using Traits = NumericAttributeTraits<T>;
+    typename Traits::StorageType storageValue;
+
+    if (isNullable)
+    {
+        DataModel::Nullable<typename Traits::WorkingType> workingValue;
```

Comment:
I need `workingValue` even after  the null check, so that I can do the range check and working2storage call.

I am unsure if there is a clean way to make this shorter/clearer.

Reply:
I guess I was envisioning a "working to storage" helper that is fallible that handles the range check and the WorkingToStorage call.

But maybe that's not worth it.

Reply:
I believe this implementation is esentially it: it does conversion to storage with range checks. If there are ideas to make the code cleaner, I would prefer as a separate update.